### PR TITLE
refactor chain randomness

### PR DESCRIPTION
**Pull request by masterjedy**
Signed-off-by: Yura Zarudniy <zarudniy@soramitsu.co.jp>

### Description of the Change
refactor chain randomness, move chain randomness calculation from chain_store to different module, so making it look unified.

### Benefits
unified appearance of randomness provider
more comfortable interface for sampling randomness.
connectivity reduced

### Possible Drawbacks 
None

### CODE REVIEW

File: core/storage/chain/impl/chain_store_impl.cpp [hunk 1]
```diff
@@ -187,58 +187,14 @@ namespace fc::storage::blockchain {
...
-        return Randomness{positive_hash};
-      }
-
-      // TODO (yuraz) I know it's very ugly, and needs to be refactored
-      // I translated it directly from go to C++
-      cids = min_ticket_block.get().parents;
-    }
+  std::shared_ptr<fc::crypto::randomness::ChainRandomnessProvider>
+  ChainStoreImpl::createRandomnessProvider() {
+    auto shared_this = shared_from_this();
```
Comment (turuslan) [inline core/storage/chain/impl/chain_store_impl.cpp]: Why not `return std::make_shared<...>(shared_from_this())`?

File: core/crypto/randomness/impl/chain_randomness_provider_impl.cpp [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+  namespace {
+    /**
+     * @brief ChainStore needs its own randomnes calculation function
+     * @return randomness value
+     */
+    crypto::randomness::Randomness drawRandomness(const Ticket &ticket,
+                                                  uint64_t round) {
+      common::Buffer buffer{};
+      const size_t bytes_required = sizeof(round) + ticket.bytes.size();
+      buffer.reserve(bytes_required);
```
Comment (turuslan) [inline core/crypto/randomness/impl/chain_randomness_provider_impl.cpp]: Why not `buffer.reserve(sizeof(round) + ticket.bytes.size())`?

File: core/crypto/randomness/chain_randomness_provider.hpp [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+   * @class ChainRandomnessProvider provides method for calculating chain
+   * randomness
+   */
+  class ChainRandomnessProvider {
+   public:
+    virtual ~ChainRandomnessProvider() = default;
+
+    /** @brief calculate randomness */
+    virtual outcome::result<Randomness> sampleRandomness(
+        const std::vector<CID> &blks, uint64_t round) = 0;
```
Comment (Alexey-N-Chernyshov) [inline core/crypto/randomness/chain_randomness_provider.hpp]: > Do not abbreviate by deleting letters within a word.
From [here](https://google.github.io/styleguide/cppguide.html#General_Naming_Rules).

File: core/crypto/randomness/impl/chain_randomness_provider_impl.hpp [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#ifndef CPP_FILECOIN_CORE_CRYPTO_RANDOMNESS_IMPL_CHAIN_RANDOMNESS_PROVIDER_IMPL_HPP
+#define CPP_FILECOIN_CORE_CRYPTO_RANDOMNESS_IMPL_CHAIN_RANDOMNESS_PROVIDER_IMPL_HPP
+
+#include "crypto/randomness/chain_randomness_provider.hpp"
+
+namespace fc::storage::blockchain {
+  class ChainStore;
```
Comment (Alexey-N-Chernyshov) [inline core/crypto/randomness/impl/chain_randomness_provider_impl.hpp]: > Avoid forward declaration
https://google.github.io/styleguide/cppguide.html#Forward_Declarations

File: core/storage/chain/impl/chain_store_impl.cpp [hunk 5]
```diff
@@ -3,56 +3,55 @@
...
-                   std::move(data_store),
-                   std::move(block_validator),
-                   std::move(weight_calculator));
-    cs.reset(new ChainStore(std::move(tmp)));
+    std::shared_ptr<ChainStoreImpl> cs{};
+    ChainStoreImpl tmp(std::move(block_store),
+                       std::move(data_store),
+                       std::move(block_validator),
+                       std::move(weight_calculator));
+    cs.reset(new ChainStoreImpl(std::move(tmp)));
```
Comment (Alexey-N-Chernyshov) [inline core/storage/chain/impl/chain_store_impl.cpp]:
```
std::shared_ptr<ChainStoreImpl> cs =
        std::make_shared<ChainStoreImpl>(std::move(block_store),
                                         std::move(data_store),
                                         std::move(block_validator),
                                         std::move(weight_calculator));
```

File: core/storage/chain/impl/chain_store_impl.hpp [hunk 6]
```diff
@@ -0,0 +1,97 @@
...
+#include "crypto/randomness/randomness_types.hpp"
+#include "primitives/cid/cid.hpp"
+#include "storage/chain/chain_data_store.hpp"
+#include "storage/chain/chain_store.hpp"
+#include "storage/ipfs/impl/ipfs_block_service.hpp"
+
+namespace fc::storage::blockchain {
+
+  /**
+   * @struct MmCids is a struct for using in mmCache cache
```
Comment (Alexey-N-Chernyshov) [inline core/storage/chain/impl/chain_store_impl.hpp]: What is mm and mmcache?

File: core/storage/chain/impl/chain_store_impl.cpp [hunk 7]
```diff
@@ -3,56 +3,55 @@
...
  */
 
-#include "storage/chain/chain_store.hpp"
+#include "storage/chain/impl/chain_store_impl.hpp"
 
-#include "common/le_encoder.hpp"
 #include "common/outcome.hpp"
-#include "datastore_key.hpp"
+#include "crypto/randomness/impl/chain_randomness_provider_impl.hpp"
+#include "storage/chain/datastore_key.hpp"
```
Comment (Alexey-N-Chernyshov) [inline core/storage/chain/impl/chain_store_impl.cpp]: Please, reorder includes and reformat code.

File: core/crypto/randomness/impl/chain_randomness_provider_impl.cpp (hunk 2 above)
Comment (masterjedy) [reply core/crypto/randomness/impl/chain_randomness_provider_impl.cpp]: for readability

File: core/crypto/randomness/chain_randomness_provider.hpp (hunk 3 above)
Comment (masterjedy) [reply core/crypto/randomness/chain_randomness_provider.hpp]: ok, renamed, but it was name from lotus implementation

File: core/storage/chain/impl/chain_store_impl.hpp (hunk 6 above)
Comment (masterjedy) [reply core/storage/chain/impl/chain_store_impl.hpp]: It is required for caching implementation,  to be used in FIL-155, I took names from lotus.

File: core/storage/chain/impl/chain_store_impl.cpp (hunk 5 above)
Comment (masterjedy) [reply core/storage/chain/impl/chain_store_impl.cpp]: This is not that simple. ChainStoreImpl has private constructor, which makes too complicated creating via std::make_shared. However, I refactored it as much as it was possible.

File: core/storage/chain/impl/chain_store_impl.hpp (hunk 6 above)
Comment (Alexey-N-Chernyshov) [reply core/storage/chain/impl/chain_store_impl.hpp]: It is better to have meaningful and descriptive names, so if you know what mmcache means, please provide description. I . managed to google only PHP-related stuff, but I hope it is not our case.

**Pull request merged**